spirv_glsl.cpp

```diff
@@ -3252,6 +3290,208 @@ string CompilerGLSL::access_chain(uint32_t base, const uint32_t *indices, uint32
+
+		return flattened_access_chain(base, indices, count, target_type, 0);
+	}
+	else
+	{
+		return access_chain(base, indices, count, false, false, need_transpose);
+	}
+}
+
+std::string CompilerGLSL::flattened_access_chain(uint32_t base, const uint32_t *indices, uint32_t count, const SPIRType &target_type, uint32_t offset)
+{
+	if (!target_type.array.empty())
+	{
+		SPIRV_CROSS_THROW("Access chains that result in an array can not be flattened");
+	}
+	else if (target_type.basetype == SPIRType::Struct)
+	{
+		return flattened_access_chain_struct(base, indices, count, target_type, offset);
+	}
+	else if (target_type.columns > 1)
+	{
+		return flattened_access_chain_matrix(base, indices, count, target_type, offset);
+	}
+	else
+	{
+		return flattened_access_chain_vector_scalar(base, indices, count, target_type, offset);
+	}
+}
+
+std::string CompilerGLSL::flattened_access_chain_struct(uint32_t base, const uint32_t *indices, uint32_t count, const SPIRType &target_type, uint32_t offset)
+{
+	std::string expr;
+
+	expr += type_to_glsl(target_type);
+	expr += "(";
+
+	for (size_t i = 0; i < target_type.member_types.size(); ++i)
+	{
+		if (i != 0)
+			expr += ", ";
+
+		const SPIRType &member_type = get<SPIRType>(target_type.member_types[i]);
+		uint32_t member_offset = type_struct_member_offset(target_type, uint32_t(i));
+
+		expr += flattened_access_chain(base, indices, count, member_type, offset + member_offset);
+	}
+
+	expr += ")";
+
+	return expr;
+}
+
+std::string CompilerGLSL::flattened_access_chain_matrix(uint32_t base, const uint32_t *indices, uint32_t count, const SPIRType &target_type, uint32_t offset)
+{
+	std::string expr;
+
+	expr += type_to_glsl(target_type);
+	expr += "(";
+
+	for (uint32_t i = 0; i < target_type.columns; ++i)
+	{
+		if (i != 0)
+			expr += ", ";
+
+		expr += flattened_access_chain_vector_scalar(base, indices, count, target_type, offset + i * 16);
+	}
+
+	expr += ")";
+
+	return expr;
+}
+
+std::string CompilerGLSL::flattened_access_chain_vector_scalar(uint32_t base, const uint32_t *indices, uint32_t count, const SPIRType &target_type, uint32_t offset)
+{
+	if (target_type.basetype != SPIRType::Float)
+		SPIRV_CROSS_THROW("Access chains that use non-floating-point base types can not be flattened");
+
+	auto result = flattened_access_chain_offset(base, indices, count, offset);
+
+	assert(result.second % 4 == 0);
+	uint32_t index = result.second / 4;
+
+	auto buffer_name = to_name(expression_type(base).self);
+
+	std::string expr;
+
+	expr += buffer_name;
+	expr += "[";
+	expr += result.first; // this is a series of N1*k1+N2*k2+... that is either empty or ends with a +
+	expr += convert_to_string(index / 4);
+	expr += "]";
+
+	expr += vector_swizzle(target_type.vecsize, index % 4);
+
+	return expr;
+}
+
+std::pair<std::string, uint32_t> CompilerGLSL::flattened_access_chain_offset(uint32_t base, const uint32_t *indices, uint32_t count, uint32_t offset, bool *need_transpose)
```

Comment:
It's actually a bit nicer that way since I can easily reuse flattened_access_chain_offset in an arbitrary location at the call site and then use the resulting expression to make the final one later.